### PR TITLE
HMAN-125: Access Control

### PR DESCRIPTION
**Pull request by kiran-yenigala-hmcts**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/HMAN-125

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X ] No
```

### CODE REVIEW

File: src/main/resources/application.yaml [hunk 1]
```diff
@@ -113,7 +113,7 @@ idam:
     totp_secret: ${IDAM_KEY_CFT_HEARING_SERVICE:AAAAAAAAAAAAAAAA}
     url: ${S2S_URL:http://localhost:4502}
   s2s-authorised:
-    services: ${CFT_HEARING_SERVICE_S2S_AUTHORISED_SERVICES:xui_webapp,hmc_hmi_inbound_adapter}
+    services: ${CFT_HEARING_SERVICE_S2S_AUTHORISED_SERVICES:xui_webapp,hmc_hmi_inbound_adapter,ccd_data}
```
Comment (danlysiak) [inline src/main/resources/application.yaml:116]:
```suggestion
    services: ${CFT_HEARING_SERVICE_S2S_AUTHORISED_SERVICES:xui_webapp,hmc_hmi_inbound_adapter}
```
😄 (Data Store won't be calling CFT Hearing Service, just the other way round)

**Review by danlysiak — COMMENTED** [review]
A few comments :)

File: src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    @Override
+    public void verifyHearingCaseAccess(Long hearingId, List<String> requiredRoles) {
+        CaseHearingRequestEntity caseHearingRequestEntity = caseHearingRequestRepository.getCaseHearing(hearingId);
+        if (caseHearingRequestEntity != null) {
+            verifyCaseAccess(caseHearingRequestEntity.getCaseReference(), requiredRoles, hearingId);
+        }
+    }
+
+    private void verifyRequiredRolesExists(List<String> requiredRoles, List<RoleAssignment> filteredRoleAssignments) {
+        if (requiredRoles.size() > 0) {
```
[the remark this answers is not in the document]
Comment (kiran-yenigala-hmcts) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java:95]: we are not checking this for non empty before, we need this check here

[the remark this answers is not in the document]
Comment (danlysiak) [reply src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java:95]: Oops, was focused on `filteredRoleAssignments` 😄

File: src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        List<RoleAssignment> filteredRoleAssignments = verifyRoleAccess(requiredRoles, hearingId);
+
+        DataStoreCaseDetails caseDetails = dataStoreRepository.findCaseByCaseIdUsingExternalApi(caseReference);
+        if (!checkRoleAssignmentMatchesCaseDetails(caseDetails, filteredRoleAssignments)) {
+            throw new InvalidRoleAssignmentException(ROLE_ASSIGNMENT_INVALID_ATTRIBUTES);
+        }
+        verifyHearingStatus(filteredRoleAssignments, caseDetails.getCaseTypeId(), hearingId);
+    }
+
+    private List<RoleAssignment> verifyRoleAccess(List<String> requiredRoles, Long hearingId) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java:83]:
```suggestion
    private List<RoleAssignment> verifyRoleAccess(List<String> requiredRoles) {
```
`hearingId` no longer used in this method

File: src/main/java/uk/gov/hmcts/reform/hmc/controllers/LinkedHearingGroupController.java [hunk 4]
```diff
@@ -16,9 +17,12 @@
 public class LinkedHearingGroupController {
 
     private LinkedHearingGroupService linkedHearingGroupService;
+    private AccessControlService accessControlService;
 
-    public LinkedHearingGroupController(LinkedHearingGroupService linkedHearingGroupService) {
+    public LinkedHearingGroupController(LinkedHearingGroupService linkedHearingGroupService,
+                                        AccessControlService accessControlService) {
         this.linkedHearingGroupService = linkedHearingGroupService;
+        this.accessControlService = accessControlService;
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/controllers/LinkedHearingGroupController.java:25]: Just need to add usage in here and `UnNotifiedHearingsController`

File: src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+        Optional<String> roleCaseType = roleAssignment.getAttributes().getCaseType();
+        if (roleCaseType == null
+            || roleCaseType.isEmpty()
+            || caseTypeId == null) {
+            return true;
+        }
+        return roleCaseType.orElse("").equals(caseTypeId);
+    }
+
+    private List<RoleAssignment> filterRoleAssignments(RoleAssignments roleAssignments) {
```
Comment (danlysiak) [inline src/main/java/uk/gov/hmcts/reform/hmc/service/AccessControlServiceImpl.java:142]: Ah missed this on previous review - we wanted to add the check that the role assignments are valid for the current time, like we do in Data Store:
https://github.com/hmcts/ccd-data-store-api/blob/e6d5579f206077c006f9ca7999ffbecca9bc89f9/src/main/java/uk/gov/hmcts/ccd/domain/model/casedataaccesscontrol/RoleAssignment.java#L39-L43

**Review by danlysiak — APPROVED** [review]

**Pull request merged**